### DISCUSSION TITLE
Report on a Terminal

### DISCUSSION THREAD

**Discussion by craft-and-code** [emmercm/igir#517] [Ideas]
Hi,

I just found something, it's not a big deal, maybe it's can be found to something else, look the result:

```
✓ Sega - 32X ··················· | 213/213 games, 3/3 BIOSes, 61/61 retail releases found
✓ Sega - Game Gear ············· | 808/808 games, 1/1 BIOSes, 490/490 retail releases found
✓ Sega - Master System - Mark II | 715/722 games, 13/13 BIOSes, 540/547 retail releases found
✓ Sega - Mega Drive - Genesis ·· | 2 799/2 801 games, 31/31 BIOSes, 1 766/1 766 retail releases found
✓ Sega - PICO ·················· | 445/445 games, 343/343 retail releases found
✓ Sega - SG-1000 ··············· | 247/247 games, 1/1 BIOSes, 239/239 retail releases found
```

`Sega - Master System - Mark II`: its `Sega - Master System - Mark III`.

I know you have to choice a 'number' to display the name of the DAT and it's great in most of cases. But can we change this for few characters?

**Comment by craft-and-code** [comment]
I profit to ask this:

I like the report in the Terminal, I use also the CSV, but I would like both.

If I use this:
`npx igir report --dat DAT --input ROM --report-output report.csv >> test.txt`

I get this in a file:

```
Sega - Master System - Mark III ...
Sega - 32X ... 213/213 games, 3/3 BIOSes, 61/61 retail releases found
Sega - Mega Drive - Genesis ...
Sega - Game Gear ... 808/808 games, 1/1 BIOSes, 490/490 retail releases found
Sega - PICO ...
Sega - Master System - Mark III ... 715/722 games, 13/13 BIOSes, 540/547 retail releases found
Sega - SG-1000 ...
Sega - Mega Drive - Genesis ... 2 799/2 801 games, 31/31 BIOSes, 1 766/1 766 retail releases found
SNK - NeoGeo Pocket ...
Sega - PICO ... 445/445 games, 343/343 retail releases found
```

I get one line empty and one line with data.

I would like the same result into the Terminal, see the first message (alphabetic order, result, etc. - and here it's good because the name of platform are complete).

Am I doing something wrong?

**Reply by emmercm** [reply]
Nope, you're not doing anything wrong here. `igir` can't print to a file the same way it does for the console, because of the updating progress bars.

For the two lines per DAT, that's because `igir` is printing when the "progress bar" is created, and then again when it's finished. When not sending the contents to a file, `igir` simply hides progress bars that had no game matches at all, to reduce spam. But I see how this is a confusing difference in behavior.

I'll change the file redirection / non-TTY behavior to match the console, so it prints the same lines. But I'm going to leave the formatting as-is, with only ASCII characters.

**Reply by craft-and-code** [reply]
I don't know what mean by:

> leave the formatting as-is, with only ASCII characters.

So I will ask something, but really I don't know what the best is: can we have an option to get the result in a Terminal also in a file? Or using this `>>` will be the job and get the Terminal log as now?

**Reply by emmercm** [reply]
`>` (and `>>`) are the best way to achieve this, as they are standard in most shells.

**Comment by craft-and-code** [comment]
I just tried to generate a file with `>>` and it's perfect, thank you.

I have 2 suggestions about it, if I may:

1. Even in a Terminal when you display igir and is version, why you don't display the web address? :)
2. It's a very small detail, but for more visibility can you think about an empty line between `Scanning for ROMs` and the list of ROMs and between this list and `Generating report ...`?

Like this:

```
 ______   ______   ______  _______  
|      \ /      \ |      \|       \ 
 \$$$$$$|  $$$$$$\ \$$$$$$| $$$$$$$\
  | $$  | $$ __\$$  | $$  | $$__| $$
  | $$  | $$|    \  | $$  | $$    $$   ROM collection manager
  | $$  | $$ \$$$$  | $$  | $$$$$$$\
 _| $$_ | $$__| $$ _| $$_ | $$  | $$   v1.8.2
|   $$ \ \$$    $$|   $$ \| $$  | $$
 \$$$$$$  \$$$$$$  \$$$$$$ \$$   \$$   https://igir.io


Scanning for DATs ... 37 unique DATs found
Scanning for ROMs ... 23 509 files found

Atari - Lynx ... 97/97 games, 1/1 BIOSes, 79/79 retail releases found
Bandai - WonderSwan Color ... 116/117 games, 2/2 BIOSes, 114/114 retail releases found
Bandai - WonderSwan ... 137/137 games, 1/1 BIOSes, 137/137 retail releases found
Bit Corporation - Gamate ... 63/64 games, 2/2 BIOSes, 63/64 retail releases found
Coleco - ColecoVision ... 200/200 games, 1/1 BIOSes, 158/158 retail releases found
GCE - Vectrex ... 45/45 games, 1/1 BIOSes, 35/35 retail releases found
Microsoft - MSX ... 948/948 games, 1/1 BIOSes, 938/938 retail releases found
Microsoft - MSX2 ... 201/201 games, 1/1 BIOSes, 196/196 retail releases found
NEC - PC Engine - TurboGrafx-16 ... 460/463 games, 8/8 BIOSes, 448/451 retail releases found
NEC - PC Engine SuperGrafx ... 5/5 games, 5/5 retail releases found
Nintendo - amiibo ... 731/731 games, 731/731 retail releases found
Nintendo - Family Computer Disk System (FDS) ... 402/402 games, 7/7 BIOSes, 374/374 retail releases found
Nintendo - Game Boy ... 2 126/2 135 games, 4/4 BIOSes, 1 672/1 672 retail releases found
Nintendo - Game Boy Advance ... 3 402/3 407 games, 3/3 BIOSes, 3 146/3 149 retail releases found
Nintendo - Game Boy Advance (Multiboot) ... 96/96 games, 95/95 retail releases found
Nintendo - Game Boy Color ... 2 011/2 011 games, 7/7 BIOSes, 1 636/1 636 retail releases found
Nintendo - Nintendo 64 (BigEndian) ... 1 106/1 107 games, 2/3 BIOSes, 972/973 retail releases found
Nintendo - Nintendo 64DD ... 22/22 games, 7/7 BIOSes, 10/10 retail releases found
Nintendo - Nintendo DSi (Digital) ... 1 068/1 068 games, 1 068/1 068 retail releases found
Nintendo - Pokemon Mini ... 44/44 games, 1/1 BIOSes, 44/44 retail releases found
Nintendo - Satellaview ... 321/321 games, 270/270 retail releases found
Nintendo - Sufami Turbo ... 13/13 games, 13/13 retail releases found
Nintendo - Super Nintendo Entertainment System ... 3 993/4 003 games, 1/1 BIOSes, 3 278/3 283 retail releases found
Nintendo - Virtual Boy ... 76/76 games, 28/28 retail releases found
Philips - Videopac+ ... 34/34 games, 2/2 BIOSes, 29/29 retail releases found
Sega - 32X ... 213/213 games, 3/3 BIOSes, 61/61 retail releases found
Sega - Game Gear ... 808/808 games, 1/1 BIOSes, 490/490 retail releases found
Sega - Master System - Mark III ... 715/722 games, 13/13 BIOSes, 540/547 retail releases found
Sega - Mega Drive - Genesis ... 2 799/2 801 games, 31/31 BIOSes, 1 766/1 766 retail releases found
Sega - PICO ... 445/445 games, 343/343 retail releases found
Sega - SG-1000 ... 247/247 games, 1/1 BIOSes, 239/239 retail releases found
SNK - NeoGeo Pocket ... 12/12 games, 2/2 BIOSes, 10/10 retail releases found
SNK - NeoGeo Pocket Color ... 128/128 games, 1/1 BIOSes, 105/105 retail releases found
VTech - CreatiVision ... 21/21 games, 1/1 BIOSes, 21/21 retail releases found
VTech - V.Smile ... 200/200 games, 198/198 retail releases found
Watara - Supervision ... 69/69 games, 69/69 retail releases found
Welback - Mega Duck ... 26/26 games, 25/25 retail releases found

Generating report ... /Users/user/Downloads/ROM/report.csv
```

Thanks, igir is get better and better :)

**Reply by emmercm** [reply]
Great suggestion, and the newlines are definitely unexpected, I will have to investigate.

**Reply by craft-and-code** [reply]
I am sorry, but I don't understand why you said it's definitely unexpected?

Just to be clear, I got this:

```
 ______   ______   ______  _______  
|      \ /      \ |      \|       \ 
 \$$$$$$|  $$$$$$\ \$$$$$$| $$$$$$$\
  | $$  | $$ __\$$  | $$  | $$__| $$
  | $$  | $$|    \  | $$  | $$    $$   ROM collection manager
  | $$  | $$ \$$$$  | $$  | $$$$$$$\
 _| $$_ | $$__| $$ _| $$_ | $$  | $$   v1.8.2
|   $$ \ \$$    $$|   $$ \| $$  | $$
 \$$$$$$  \$$$$$$  \$$$$$$ \$$   \$$


Scanning for DATs ... 37 unique DATs found
Scanning for ROMs ... 23 509 files found
Atari - Lynx ... 97/97 games, 1/1 BIOSes, 79/79 retail releases found
... ROMS ROMS ROMS
Welback - Mega Duck ... 26/26 games, 25/25 retail releases found
Generating report ... /Users/user/Downloads/ROM/report.csv
```

And I think it will be better like this:

```
 ______   ______   ______  _______  
|      \ /      \ |      \|       \ 
 \$$$$$$|  $$$$$$\ \$$$$$$| $$$$$$$\
  | $$  | $$ __\$$  | $$  | $$__| $$
  | $$  | $$|    \  | $$  | $$    $$   ROM collection manager
  | $$  | $$ \$$$$  | $$  | $$$$$$$\
 _| $$_ | $$__| $$ _| $$_ | $$  | $$   v1.8.2
|   $$ \ \$$    $$|   $$ \| $$  | $$
 \$$$$$$  \$$$$$$  \$$$$$$ \$$   \$$   https://igir.io


Scanning for DATs ... 37 unique DATs found
Scanning for ROMs ... 23 509 files found

Atari - Lynx ... 97/97 games, 1/1 BIOSes, 79/79 retail releases found
... ROMS ROMS ROMS
Welback - Mega Duck ... 26/26 games, 25/25 retail releases found

Generating report ... /Users/user/Downloads/ROM/report.csv
```

Thanks :)

**Reply by emmercm** [reply]
Ah, I misunderstood, I thought you meant that you were getting the double newlines, not that you are requesting them.

Can I ask what your use case is for redirected output? This output is not supposed to be pretty, it's really only useful for generating debug logs that can be attached to submitted bug reports or feature requests.

**Reply by craft-and-code** [reply]
OK no problem :)

I like to use this output because:
- I got numbers, so I can see how many ROMs make the set
- Which rom set I have
- **More importantly**, I can see how many are missing (and maybe if I have a problem with some? - never happen so I am not sure) -> _it's good with Terminal because you got colours, which you don't have in a file_.

I like the CSV report too, but it's handy when you need more specific details but need to be open with software.

Here you can see the minimal information, quickly, into a Terminal or just with a text editor.
No need to make it more beautiful, new lines are just to get more readability.

**Comment by craft-and-code** [comment]
I little thinking about colours in Terminal.

Colours are good, but I can see 2 problems:
1. Some people, like me, can have some issues with colours (I have a problem seeing the yellow with igir)
2. Colours can be change, I mean in a settings into the Terminal

What about keeping colours but with something more, like bold with numbers when you have some missing ROMs?

Just a thought, I know your answer about reports in a Terminal :).

**Reply by emmercm** [reply]
For coloring of the number counts, I'm specifying RGB values explicitly:

https://github.com/emmercm/igir/blob/5254ad42e94b282effd0176deadd516313907715/src/types/datStatus.ts#L130-L142

so I'm hoping that minimizes differences between terminals. But the library I'm using will downsample and approximate the color for terminals that only support ANSI colors:

https://github.com/chalk/chalk#256-and-truecolor-color-support

I found it fairly difficult to pick obviously different green and yellow colors. I'm not sure bolding will help solve, I personally find bold terminal characters hard to differentiate as well.

**Reply by craft-and-code** [reply]
OK for the bold thing.

What do you think about this?

```
 ______   ______   ______  _______  
|      \ /      \ |      \|       \ 
 \$$$$$$|  $$$$$$\ \$$$$$$| $$$$$$$\
  | $$  | $$|    \  | $$  | $$    $$   ROM collection manager
  | $$  | $$|    \  | $$  | $$    $$   https://igir.io/
  | $$  | $$ \$$$$  | $$  | $$$$$$$\
 _| $$_ | $$__| $$ _| $$_ | $$  | $$   v1.9.0
|   $$ \ \$$    $$|   $$ \| $$  | $$
 \$$$$$$  \$$$$$$  \$$$$$$ \$$   \$$


✓ Scanning for DATs ············ | 37 unique DATs found
✓ Scanning for ROMs ············ | 23 509 files found

✓ Atari - Lynx ················· | 97/97 games, 1/1 BIOSes, 79/79 retail releases found
x Bandai - WonderSwan Color ···· | 116/117 games, 2/2 BIOSes, 114/114 retail releases found

✓ Generating report ············ | /Users/sebr_d/Downloads/ROM/report.csv
```

My first idea was like this:

```
 ______   ______   ______  _______  
|      \ /      \ |      \|       \ 
 \$$$$$$|  $$$$$$\ \$$$$$$| $$$$$$$\
  | $$  | $$|    \  | $$  | $$    $$   ROM collection manager
  | $$  | $$|    \  | $$  | $$    $$   https://igir.io/
  | $$  | $$ \$$$$  | $$  | $$$$$$$\
 _| $$_ | $$__| $$ _| $$_ | $$  | $$   v1.9.0
|   $$ \ \$$    $$|   $$ \| $$  | $$
 \$$$$$$  \$$$$$$  \$$$$$$ \$$   \$$


✓ Scanning for DATs ············ | 37 unique DATs found
✓ Scanning for ROMs ············ | 23 509 files found

✓ Atari - Lynx ················· | ✓ 97/97 games, 1/1 BIOSes, 79/79 retail releases found
✓ Bandai - WonderSwan Color ···· | x 116/117 games, 2/2 BIOSes, 114/114 retail releases found

✓ Generating report ············ | /Users/sebr_d/Downloads/ROM/report.csv
```

But you use it already in first position, and I think in some situation you display a mark icon? It's just an idea.
Maybe using a different icon when your ROM set it's not completed instead the mark icon?

**Reply by emmercm** [reply]
I'd be worried about the X making people think the write failed, rather than the set being incomplete. So I think your first idea might be more clear.

Or maybe there's a good way to include a percentage? I think a "100%" would be easy to spot at a glance.

**Reply by craft-and-code** [reply]
> I'd be worried about the X making people think the write failed, rather than the set being incomplete. So I think your first idea might be more clear.

Yes and I think you are right.

> Or maybe there's a good way to include a percentage? I think a "100%" would be easy to spot at a glance.

I just checked emojis... So "100%" might be the best :)
